### ISSUE TITLE
Angular 12 error

### ISSUE THREAD

**Issue by quangquy87** [rxweb/rxweb#490]
# Describe the bug

rror: node_modules/@rxweb/reactive-form-validators/services/rx-form-array.d.ts:28:5 - error TS2416: Property 'removeAt' in type 'RxFormArray' is not assignable to the same property in base type 'FormArray'.  
  Type '(index: number, isRemovedInstance?: boolean) => void' is not assignable to type '(index: number, options?: { emitEvent?: boolean; }) => void'.
    Types of parameters 'isRemovedInstance' and 'options' are incompatible.
      Type '{ emitEvent?: boolean; }' is not assignable to type 'boolean'.

28     removeAt(index: number, isRemovedInstance?: boolean): void;
       ~~~~~~~~


# Expected Behavior




# To Reproduce



# Url (stackblitz or plunker or other)



# Package Version
@rxweb/<package-name>@<packageversion>

**Comment by Barzille** [comment]
Same for me here, any way to change this on our side?

**Comment by brane53** [comment]
I'm having the same issue, but I'm also getting an error for the `push` method as well.

It appears in Angular v12 the API for several methods of [FormArray](https://angular.io/api/forms/FormArray) and [FormGroup](https://angular.io/api/forms/FormGroup) classes have changed slightly by adding an options argument that wasn't there in v11. See [Breaking Changes](https://github.com/angular/angular/blob/master/CHANGELOG.md#breaking-changes) section for more on which methods are affected.

rxweb does in fact extend these methods based on the old API, so I think it's just a matter of changing those methods to include the new options argument. Just as a visual here's the code for extending `removeAt`

``` typescript
removeAt(index: number, isRemovedInstance: boolean = false) {
    let formGroup: any = this.root;
    if (!isRemovedInstance)
        this.arrayObject.splice(index, 1);
    else {
        for (var i = index; i < this.arrayObject.length - 1; i++)
            this.arrayObject[i] = this.arrayObject[i + 1];
        this.arrayObject.pop();
    }
    super.removeAt(index);
    if (formGroup[VALUE_CHANGED_SYNC])
        formGroup.valueChangedSync();
    this.patch();
    this.checkValidation();
}
```

I'm thinking the change should be

``` typescript
removeAt(index: number, options: { emitEvent?: boolean } = {}, isRemovedInstance: boolean = false) {
    ...
    super.removeAt(index, options);
    ...
}
```
 
There may be more to do, but I think this is the core of the problem.

**Comment by chobijaeyu** [comment]
same error.
```
Error: node_modules/@rxweb/reactive-form-validators/services/rx-form-array.d.ts:14:5 - error TS2416: Property 'push' in type 'RxFormArray' is not assignable to the same property in base type 'FormArray'.
  Type '(control: any, isAddedInstance?: boolean) => void' is not assignable to type '(control: AbstractControl, options?: { emitEvent?: boolean; }) => void'.
    Types of parameters 'isAddedInstance' and 'options' are incompatible.
      Type '{ emitEvent?: boolean; }' is not assignable to type 'boolean'.

14     push(control: any, isAddedInstance?: boolean): void;
       ~~~~


Error: node_modules/@rxweb/reactive-form-validators/services/rx-form-array.d.ts:28:5 - error TS2416: Property 'removeAt' in type 'RxFormArray' is not assignable to the same property in base type 'FormArray'.
  Type '(index: number, isRemovedInstance?: boolean) => void' is not assignable to type '(index: number, options?: { emitEvent?: boolean; }) => void'.
    Types of parameters 'isRemovedInstance' and 'options' are incompatible.
      Type '{ emitEvent?: boolean; }' is not assignable to type 'boolean'.

28     removeAt(index: number, isRemovedInstance?: boolean): void;
```


Considering the current situation in India, I hope everything is fine for him.stay safe.

**Comment by pl4yradam** [comment]
+ 1 here

**Comment by olavaukan** [comment]
Having the same problem. This blocks my upgrade to Angular 12.

I can't figure out how to build the source (missing tsconfig.app.json?) and the community guidelines for submitting pull requests return 404, so I don't know the "correct" way to do this but messing around with the files in node_modules I can make the error go away locally. 

Like @brane53 says, the issue is with the push and removeAt methods. I've changed the signature in these two methods in my `rx-form-array.d.ts` file:
```
import { FormArray } from "@angular/forms";
import { ResetFormType } from "../enums/reset-type";
export declare class RxFormArray extends FormArray {
    private arrayObject;
    private arrayConfig?;
    private _baseValue;
    private _isModified;
    private _modified;
    constructor(arrayObject: any[], controls: any, validatorOrOpts?: any, asyncValidator?: any, arrayConfig?: {
        allowMaxIndex?: number;
        messageKey?: string;
    });
    readonly isModified: boolean;
    push(control: any, options?: { emitEvent?: boolean; }, isAddedInstance?: boolean): void;
    patch(): void;
    resetForm(options?: {
        index: number;
        groupOption: {
            resetType?: ResetFormType;
            with?: string[];
            value?: {
                [key: string]: any;
            };
        };
        pushFunction: (value: Object) => boolean;
    }): void;
    commit(): void;
    removeAt(index: number, options?: { emitEvent?: boolean; }, isRemovedInstance?: boolean): void;
    private checkValidation;
    private checkModification;
    private cloneObject;
}
```
I've changed the methods in my (esm205) `rx-form-array.js` file like this:

```
    push(control, options, isAddedInstance = false) {
        let formGroup = this.root;
        if (this.arrayObject)
            if (control.modelInstance) {
                if (!isAddedInstance)
                    this.arrayObject.push(control.modelInstance);
                else
                    this.arrayObject[this.arrayObject.length] = control.modelInstance;
            }
        super.push(control, options);
        if (formGroup[VALUE_CHANGED_SYNC])
            formGroup.valueChangedSync();
        this.patch();
        this.checkValidation();
    }

    removeAt(index, options, isRemovedInstance = false) {
        let formGroup = this.root;
        if (!isRemovedInstance)
            this.arrayObject.splice(index, 1);
        else {
            for (var i = index; i < this.arrayObject.length - 1; i++)
                this.arrayObject[i] = this.arrayObject[i + 1];
            this.arrayObject.pop();
        }
        super.removeAt(index, options);
        if (formGroup[VALUE_CHANGED_SYNC])
            formGroup.valueChangedSync();
        this.patch();
        this.checkValidation();
    }
```
And similarly in my (esm5) `rx-form-array.js` file like this:
```
    RxFormArray.prototype.push = function (control, options, isAddedInstance) {
        if (isAddedInstance === void 0) { isAddedInstance = false; }
        var formGroup = this.root;
        if (this.arrayObject)
            if (control.modelInstance) {
                if (!isAddedInstance)
                    this.arrayObject.push(control.modelInstance);
                else
                    this.arrayObject[this.arrayObject.length] = control.modelInstance;
            }
        _super.prototype.push.call(this, control, options);
        if (formGroup[VALUE_CHANGED_SYNC])
            formGroup.valueChangedSync();
        this.patch();
        this.checkValidation();
    };

    RxFormArray.prototype.removeAt = function (index, options, isRemovedInstance) {
        if (isRemovedInstance === void 0) { isRemovedInstance = false; }
        var formGroup = this.root;
        if (!isRemovedInstance)
            this.arrayObject.splice(index, 1);
        else {
            for (var i = index; i < this.arrayObject.length - 1; i++)
                this.arrayObject[i] = this.arrayObject[i + 1];
            this.arrayObject.pop();
        }
        _super.prototype.removeAt.call(this, index, options);
        if (formGroup[VALUE_CHANGED_SYNC])
            formGroup.valueChangedSync();
        this.patch();
        this.checkValidation();
    };
```
No clue if this breaks everything and destroys the universe as we know it, but I think the fix is something along these lines...

**Comment by ajayojha** [comment]
this has been fixed in the new release of 2.1.3, kindly update the same.
